sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
-            valueInGlobalWindow("C"),
-            valueInGlobalWindow("D"),
-            valueInGlobalWindow("E"),
-            valueInGlobalWindow("F")));
+    @Parameterized.Parameter(0)
+    public ProcessBundleSplitResponse expectedResponse;
+
+    @Parameterized.Parameter(1)
+    public long inputElements;
+
+    @Parameterized.Parameter(2)
+    public List<String> processedElements;
+
+    @Parameterized.Parameter(3)
+    public double fractionOfRemainder;
+
+    @Parameterized.Parameter(4)
+    public double elementProgress;
+
+    @Test
+    public void testElementSplit() throws Exception {
+      SplittingReceiver splittingReceiver = mock(SplittingReceiver.class);
+      BeamFnDataClient mockBeamFnDataClient = mock(BeamFnDataClient.class);
+      when(splittingReceiver.getProgress()).thenReturn(elementProgress);
+      when(splittingReceiver.trySplit(anyDouble())).thenReturn(mockSplitResult);
+      BeamFnDataReadRunner<String> readRunner =
+          createReadRunner(splittingReceiver, PTRANSFORM_ID, mockBeamFnDataClient);
+
+      assertEquals(
+          expectedResponse,
+          executeSplit(
+              readRunner,
+              PTRANSFORM_ID,
+              inputElements,
+              processedElements,
+              fractionOfRemainder,
+              Collections.EMPTY_LIST));
+    }
   }
 
-  @Test
-  public void testSplittingDownstreamReceiver() throws Exception {
-    SplitResult splitResult =
-        SplitResult.of(
-            BundleApplication.newBuilder().setInputId("primary").build(),
-            DelayedBundleApplication.newBuilder()
-                .setApplication(BundleApplication.newBuilder().setInputId("residual").build())
-                .build());
-    SplittingReceiver splittingReceiver = mock(SplittingReceiver.class);
-    when(splittingReceiver.getProgress()).thenReturn(0.3);
-    when(splittingReceiver.trySplit(anyDouble())).thenReturn(splitResult);
-    BeamFnDataReadRunner<String> readRunner = createReadRunner(splittingReceiver);
-
-    ProcessBundleSplitRequest request =
-        ProcessBundleSplitRequest.newBuilder()
-            .putDesiredSplits(
-                "pTransformId",
-                DesiredSplit.newBuilder()
-                    .setEstimatedInputElements(10)
-                    .setFractionOfRemainder(0.05)
-                    .build())
-            .build();
-    ProcessBundleSplitResponse.Builder responseBuilder = ProcessBundleSplitResponse.newBuilder();
-
-    // We will be "processing" the 'C' element, aka 2nd index
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("A"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("B"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("C"));
-    readRunner.trySplit(request, responseBuilder);
+  // Test different cases of element split with non-empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ElementSplitWithAllowedSplitPointsTest {
```

Comment:
The same to [test_sdf_split_with_allowed_splits](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor_test.py#L124)